File: 백준/Gold/10159. 저울/저울.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static int M;
    static ArrayList<ArrayList<Integer>> front;
    static ArrayList<ArrayList<Integer>> back;
    static boolean[] visited;
    static int cnt;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        N = Integer.parseInt(br.readLine());
        M = Integer.parseInt(br.readLine());

        front = new ArrayList<>();
        back = new ArrayList<>();
        for (int i = 0; i < N+1; i++) {
            front.add(new ArrayList<>());
            back.add(new ArrayList<>());
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int f = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            front.get(b).add(f);
            back.get(f).add(b);
        }

        for (int i = 1; i < N+1; i++) {
            visited = new boolean[N+1];
            cnt = 1;
            searchBack(i, i);
            searchFront(i, i);
            sb.append(N-cnt).append("\n");
        }

        System.out.println(sb);
    }

    static void searchFront(int root, int n) {

        visited[n] = true;

        if (root != n) {
            cnt++;
        }

        for (int i : front.get(n)) {
            if (!visited[i]) {
                searchFront(root, i);
            }
        }
    }

    static void searchBack(int root, int n) {
        visited[n] = true;

        if (root != n) {
            cnt++;
        }

        for (int i: back.get(n)) {
            if (!visited[i]) {
                searchBack(root, i);
            }
        }
    }
}

File: 백준/Gold/10159. 저울/저울.py
# 저울
import sys
input = sys.stdin.readline

def search(arrlist, root, n):
    global cnt
    
    visited[n] = True

    if root != n:
        cnt += 1

    for i in arrlist[n]:
        if visited[i] == 0:
            search(arrlist, root, i)
            
N = int(input())
M = int(input())

front = [[] for _ in range(N+1)]
back = [[] for _ in range(N+1)]

for i in range(M):
    f, b = map(int,input().split())
    front[b].append(f)
    back[f].append(b)

for i in range(1, N+1):
    visited = [0] * (N+1)
    cnt = 1
    search(front, i, i)
    search(back, i, i)
    print(N-cnt)
